### ISSUE TITLE
WebGL Primitives Discussion

### ISSUE THREAD

**Issue by Siphonophora** [processing/p5.js#2226]
Hi, I am new to this project and have been working on one of the issues @kjhollen created to add features to WebGL. After looking through the implementation, I wanted to sum what my understanding of what exists and my thoughts on what I could help with. My main assumption is that we would want users to generally be able to turn on WebGL without breaking a sketch, and further, to be easily able to substitute 3d primitives for 2d ones. 

I would be very interested in hearing feedback on this, what are good ideas and what wouldn't make sense for p5.

**2D Primitives** 
These primitives all inherit their documentation, argument validation, etc from the core 2D primitives. Ideally, a user should be able to enable WEBGL for an existing sketch with these primitives and have it continue to work (the only exception being that (0,0) is now in the center of the frame. 
- All take x and y as their first two arguments
- These primitives always draw at z=0. 
- Arguments should only be added in WEBGL, not altered. 
-- Rect in WebGL takes 6 arguments instead of 8, with the last two being different than the standard args. I can fix this so the WebGL rect can handle curved corners.
- Detail arguments should only control the number of vertices at the edge of the primitive, and should not tessellate the face of the primitive
--Detail needs to consistently be part of the gID for these shapes, which isn't consistently true.
- Stroke should draw around the edge of the primitive and should not cross the face (a few inconsistencies I can fix there)
-- Line and the curves don't obey stroke. but also are not really 2d primitives so out of scope for what I am thinking about
- Attributes
--only strokeWeight() is implemented. At high weights, it can give a these primitives some apparent depth. 
-- ellipseMode() and rectMode() need implementation, which I could work on. 
-- noSmooth(), smooth(), strokeCap(), strokeJoin() need implementation / or friendly errors, as they crash the sketch now. 
-Texturing. I think we should have a minimum two texture modes
--Stretched (the current behavior) - arc and ellipse would place the top of the image at their centers. Rect won't look different, but quad would assign each vertex to a corner of the texture. Probably this isn't often what users would want. Triangle would be mapped to 3 specific points on the texture. Below is a 5 sided ellipse on the left and  quad on the right.
![stretchedtextures](https://user-images.githubusercontent.com/32316111/31049277-17c9448e-a5fe-11e7-8fd7-0aee20e7aacf.png)
--Standard (new default) - Take the primitive, determine the smallest rectangle it can fit in and scale that to the texture.
![standardmappedtextures](https://user-images.githubusercontent.com/32316111/31049315-e0169e6e-a5fe-11e7-8273-e220bf7464ab.png)
 
**3D Primitives**
 These should be entirely new primitives. 
- All _should_ take x, y, z as their first arguments. I think it is strange that this isn't true. The convention for the 2d primitives is consistently to take first the _where_ arguments and then the arguments defining the shape. The current implementation introduces a lot of extraneous push, translate, pop operations and would make it harder to convert a 2d sketch to 3d.
--Perhaps some different mode to switch this on and preserve the old api?
- plane() should be deprecated/removed in favor of rect() unless something is done to make it distinct from rect(). 
- sphere's existence is inconsistent with the 2D primitives. We don't supply a circle primitive and shouldn't offer sphere. Ellipsoid can completely cover that functionality, just as ellipse does.  
--Presumably api breaking changes are not the best, but would there be a problem with turning Sphere into a wrapper for ellipsoid? We have nearly identical code for both. For this and the suggestion below, some consolidation of the code would make adding extra texture features easier. 
- cone and cylinder should be consolidated into a cylinder which can take 4 different radius arguments (x and y) for top and bottom. This would allow for cylinders, cones and tapered cylinders all to be drawn. Perhaps either cap of the cylinder can be optional. 
- Detail should only be offered for curved surfaces. So cylinder would get one detail argument while ellipsoid and torus would get two. 
--The documentation should emphasize the visual impact of reducing the detail. It adds a lot of flexibility to understand that the cylinder is really an extruded ngon. 
![cylinderxdetail](https://user-images.githubusercontent.com/32316111/31049078-a3db587c-a5f9-11e7-8431-bbe4a9357e2e.png)
--Documentation for ellipsoid should explain the visual artifacts you can get based on a lower detail Y with textures. (below is a view of the pole of a sphere with default detail on the left, higher on the right)
![spheredetailartifiacts](https://user-images.githubusercontent.com/32316111/31049034-d736aec0-a5f8-11e7-92f5-3c6d82b3e4f5.png)
-Documentation, in general, for these primitives needs to be fixed. Nearly all are drawn larger than their canvases. I saw some issues are/were outstanding for the documentation, so perhaps this change is already coming?
- For stroke, I would assume the desired effect would be somewhat like cell shading. Stroke should not be drawn on an edge if the attached faces are on the same plane, within some tolerance ( < 20 degrees would keep stroke from being drawn on a smooth ellipsoid or on the sides of a smooth cylinder, but the 'rims' of the cylinder would get edges).
- Texture
-- ellipsoid and torus are easy, as they just map their whole surfaces onto the texture. Perhaps offer tiling of the texture on x and y. 
-- Box and cylinder can be fit into a single model for texturing. A texture map can have a top, waist, and bottom. Below, the green would be stretched around the side of the cylinder, while the purple would be applied to the bottom. Perhaps some helper function could assemble this file from multiple files. 
![texturemap](https://user-images.githubusercontent.com/32316111/31048977-cfcf18a8-a5f7-11e7-880e-cf34cf0f6007.png)
---Box could also get a standard unwrapped box option.

Let me know what you think!

**Comment by kjhollen** [comment]
hi, thanks for sharing this! I'm still considering the details in this post, but I think one thing that might be helpful is [this document](https://docs.google.com/document/d/1ZT27SS_NmVv4c7-2x9_H44xQ7UNmf1wuTpcAraHmB7c/edit?usp=sharing) that we put together for Google Summer of Code—I'm realizing that I linked this from my write-up but it's not linked from the project directly yet. I'll figure out where it can go.

The main set of goals/tasks we put together were:

- 2D mode feature parity
- implementing stroke and fill
- performance
- better control of geometry cache
- porting features from Processing's 3D mode
- documentation for contributors

One of the things we considered was how sophisticated the 3D library ought to be. We were thinking of it as something like a stepping stone for three.js—where someone might practice the basic ideas of 3D graphics while they're learning, but for heavier projects they'd move on to another toolkit.

The existing API for 3D shapes matches the Processing implementation. It's nice to match the feature set to allow porting of Processing projects (and to get the API close, too) but should also make adjustments for new Web GL conventions. A quick look at the three.js API ([like this box example](https://threejs.org/docs/index.html#api/geometries/BoxGeometry)) suggests that it might still be common to define a 3D shape by its dimensions and to consider any position/rotation movement as a transform. As a teacher, I think the API as implemented now is well poised to help students start thinking about these concepts. I can see it being easier to use but also it would miss some teaching opportunities.

in any case, I think the other suggestions (like sharing code between ellipsoid and sphere, and cone/cylinder) are great improvements that can be moved on right away, regardless of API change. anything to make the code simpler and easier to maintain!

**Comment by mlarghydracept** [comment]
Hi! This is great! Thanks for looking into this in so much depth.

I don't have time at the moment to respond to each point but I just wanted to echo that code sharing between those primitives would be very welcome! But I am not sure that I agree that cone and sphere should be eliminated. I understand that similar effects could be achieved with cylinder and ellipsoid respectively, but I am not sure if it is useful to ask users to customize another shape to achieve a different fundamental shape. Quick and easy access to the sphere and cone feels useful enough and not really cumbersome for the library. 

I also just would echo @kjhollen's sentiment about 3D shapes matching the Processing implementation. I think the current degree of similarity between 3D shapes in p5 and Processing is a strength.

As far as diagonal stroking across faces, I am not sure what is preferable here. The wireframe look here is probably desirable to some users and not others. I did manually remove the diagonal strokes from the box primitive, which is potentially visually inconsistent with the other shapes. Maybe it should be possible to switch between a cel-shaded and a more traditional wireframe. Not sure!

Anyways thanks again for the research, excited to see progress!

**Comment by Siphonophora** [comment]
Thanks for the feedback. I will start with 2D/3D parity before other items. Here is what I have for a punch list. 

- [x] Rect needs rounded corners (Done, except for documentation pending the resolution to #2229). Incidentally, it would not be difficult to add this feature to quad.

- [ ] Add a manual test to use to verify parity of the 2d primitives in 2D and 3D. Excepting once point is done, we should be all set. 

- [ ] Add a function to p5.Geometry to validate the inputs. Presumably just for developers, but its fairly difficult as is to tell what is wrong when you pass bad parameters

- [ ] Add strokeIndicies to p5.Geometry prototype, it seems to be the only array that it used within geometry which is not created in the constructor. 

- [ ] Line is drawn using create shape and says it doesn't use stroke, but fill to set color. Might just be comments which are wrong


@mlarghydracept regarding different preferences for strokes. We could add a new function to attributes like strokeMesh(). Args could be DEFAULT (minimal strokes like you set for the box) and WIREFRAME. _makeTriangleEdges in geometry could check the property and if it is set to wireframe, ignore strokeIndicies.

One followup on the earlier items, I realized that adding x,y,z to the 3d shapes probably is a bad idea we start putting them in the geometry hashes.

**Comment by Siphonophora** [comment]
I updated the checklist above. I have also done a bit on simplifying the code for 3d primitives and playing with options like 'wireframe' above. As I am working on these, is there any value to having all the horizontal divisions in the cylinder? If we are offering something pretty simple without the ability to deform the primitives, I would guess it is only costing us in rendering time. 
![image](https://user-images.githubusercontent.com/32316111/31363579-124495dc-ad2d-11e7-8ddd-54b104b467f9.png)


Below is the manual test I added to help verify 2d and Web gl are interchangeable. 

![image](https://user-images.githubusercontent.com/32316111/31363364-e3d09df0-ad2b-11e7-84c6-e29d737c91b2.png)

I am not too sure as this progresses if these should be broken into different PRs or done as a group. I can see the test above being easy to add, but the other code taking more work to review/approve. Any suggestions on the right way to balance that?

**Comment by Siphonophora** [comment]
Also, any thoughts on why noFill() doesn't seem to do anything?

**Comment by mlarghydracept** [comment]
Hey! Thanks for all the work! Sorry for the slow response. I am a bit swamped with other work at the moment. 

I am not sure there is any value to having the horizontal stroke in cylinder. Maybe this could only happen in WIREFRAME mode? I think taking liberties with what is considered an outside edge in the DEFAULT mode makes sense and will have a nice visual outcome.

Re noFill(), I don't have the opportunity to check right now. What you mean by it doesn't seem to do anything? It should cause the renderer to  ignore drawing the faces. Have you checked the wireframe example in the manual tests folder?

**Comment by kjhollen** [comment]
thank you for taking on this work, @Siphonophora! I think it is worth opening separate issues/PRs for work items as they materialize (note that there are many open issues for missing 3D features already). let's leave this larger issue open for another week or so to see if anyone else would like to provide feedback, then we can close an open specific items for issues not covered in current open tickets.

I think I am in favor of the cone and cylinder possibly being the same implementation underneath (but keeping both as separate API for the user), but that's the only reason I can think of for keeping the horizontal layers in the cylinder right now.

**Comment by kjhollen** [comment]
Closing this for organizational purposes, because no additional comments have been added since October. @Siphonophora would you like to open a more specific issue for the remaining work item here (validation of geometry)?

**Comment by micuat** [comment]
@Siphonophora @kjhollen I just got smooth not defined bug and found this thread. I think, at least for now, there should be stub functions that says it's not available in WEBGL; otherwise it may appear as a mysterious bug (for example when you switch a sketch from p2d to webgl).

**Comment by kjhollen** [comment]
thanks @micuat! I think most of the unimplemented primitive shape functions do print errors like you suggest, but it seems that this isn't implemented for smooth(). Can you please file a separate issue describing what's happening, so that we can track it independently of this discussion thread? thank you!
